src/LinuxTracing/PerfEventReaders.cpp

```diff
@@ -18,12 +18,119 @@
+    ring_buffer->ReadRawAtOffset(&event.addr, current_offset, sizeof(uint64_t));
+    current_offset += sizeof(uint64_t);
+  }
+
+  if (flags.sample_type & PERF_SAMPLE_ID) {
+    ring_buffer->ReadRawAtOffset(&event.id, current_offset, sizeof(uint64_t));
+    current_offset += sizeof(uint64_t);
+  }
+
+  if (flags.sample_type & PERF_SAMPLE_STREAM_ID) {
+    ring_buffer->ReadRawAtOffset(&event.stream_id, current_offset, sizeof(uint64_t));
+    current_offset += sizeof(uint64_t);
+  }
+
+  if (flags.sample_type & PERF_SAMPLE_CPU) {
+    ring_buffer->ReadRawAtOffset(&event.cpu, current_offset, sizeof(uint32_t));
+    current_offset += sizeof(uint32_t);
+    ring_buffer->ReadRawAtOffset(&event.res, current_offset, sizeof(uint32_t));
+    current_offset += sizeof(uint32_t);
+  }
+
+  if (flags.sample_type & PERF_SAMPLE_PERIOD) {
+    ring_buffer->ReadRawAtOffset(&event.period, current_offset, sizeof(uint64_t));
+    current_offset += sizeof(uint64_t);
+  }
+
+  if (flags.sample_type & PERF_SAMPLE_CALLCHAIN) {
+    ring_buffer->ReadRawAtOffset(&event.ips_size, current_offset, sizeof(uint64_t));
+    current_offset += sizeof(uint64_t);
+    event.ips = make_unique_for_overwrite<uint64_t[]>(event.ips_size);
+    ring_buffer->ReadRawAtOffset(event.ips.get(), current_offset,
+                                 event.ips_size * sizeof(uint64_t));
+    current_offset += event.ips_size * sizeof(uint64_t);
+  }
+
+  if (flags.sample_type & PERF_SAMPLE_RAW) {
+    ring_buffer->ReadRawAtOffset(&event.raw_size, current_offset, sizeof(uint32_t));
+    current_offset += sizeof(uint32_t);
+    event.raw_data = make_unique_for_overwrite<uint8_t[]>(event.raw_size);
+    ring_buffer->ReadRawAtOffset(event.raw_data.get(), current_offset,
+                                 event.raw_size * sizeof(uint8_t));
+    current_offset += event.raw_size * sizeof(uint8_t);
+  }
+
+  if (flags.sample_type & PERF_SAMPLE_REGS_USER) {
+    ring_buffer->ReadRawAtOffset(&event.abi, current_offset, sizeof(uint64_t));
+    current_offset += sizeof(uint64_t);
+    if (event.abi != PERF_SAMPLE_REGS_ABI_NONE) {
+      const int num_of_regs = std::bitset<64>(flags.sample_regs_user).count();
+      event.regs = make_unique_for_overwrite<uint64_t[]>(num_of_regs);
+      ring_buffer->ReadRawAtOffset(event.regs.get(), current_offset,
+                                   num_of_regs * sizeof(uint64_t));
+      current_offset += num_of_regs * sizeof(uint64_t);
+    }
+  }
+
+  if (flags.sample_type & PERF_SAMPLE_STACK_USER) {
+    ring_buffer->ReadRawAtOffset(&event.stack_size, current_offset, sizeof(uint64_t));
+    current_offset += sizeof(uint64_t);
+    event.stack_data = make_unique_for_overwrite<uint8_t[]>(event.stack_size);
+    ring_buffer->ReadRawAtOffset(event.stack_data.get(), current_offset,
+                                 event.stack_size * sizeof(uint8_t));
+    current_offset += event.stack_size * sizeof(uint8_t);
+    if (event.stack_size != 0u) {
```

Comment:
do we want to have a similar safety check for the "ips"?

Reply:
No, this is only relevant to the stack. perf event man page mentions that if `size == 0` then dyn_size won't be given. That might happen if you try to profile kernel processes.

Reply:
Ah so this is about the dynsize. I was wandering about the "data" field. But it make sense (also we will move the offset by "ips_size" which is 0, so all is fine).